### DISCUSSION TITLE
C# 12: Collection literals

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#65] [General]
# C# 12: Collection literals

A new feature is hitting C# 12 and it is called Collection literals. In this blog post, I will show you what it is and how it works.

https://steven-giesel.com/blogPost/9916410c-0428-4caf-b88e-9fbae72b6a1f

**Comment by dimohy** [comment]
var items = [1, 2, 3];

Future C# developer candidates will want the clean code above. Is this really possible, or has C# already come a long way?

**Reply by linkdotnet** [reply]
The propsal intends the usage you describe above. Depending hin the context it should choose the „best“ option possible.

**Reply by KennethHoff** [reply]
Generally speaking, `List` is the "standard collection" in C#, so `var items = [1, 2, 3];` would be equal to `List<int> items = [1, 2, 3];` which itself would be more-or-less the same as `List<int> items = new List<int>() {1, 2, 3};`.
IIRC there are differences between the new way(= [1,2,3]) and the old way( =new List<....)  in terms of how it works though - the new one is more performant I believe

**Reply by linkdotnet** [reply]
In its current stage, it does get lowered to somewhat the same code. But the proposal goes a bit further. Just imagine this:
```csharp
private void MyFunc()
{
  var a = [1, 2, 3];
  // Do something with a
}
```

a could be either of type `Span<int>` or just `int[]` as this is more performant. I do like this though.

**Reply by KennethHoff** [reply]
I realize now that my comment was very affirmative - pretending like I know what I'm talking about - but after posting it I did some more reading up on the WG drafts and it does seem like there is no defined natural type for this. The only definite thing was that `var a = []` will not be allowed in the first version.

It would be really cool if they used Spans in scenarios where they've concluded that it could - Free performance!
